Events and logs api (#2676)
open-telemetry/opentelemetry-specification: specification/logs/semantic_conventions/events.md

# Semantic Convention for event attributes

**Status**: [Experimental](../../document-status.md)

> This document describes the attributes of standalone Events that are represented by `LogRecord`s. All standalone Events have a name and a domain. The Event domain is used as a mechanism to avoid conflicts with event names.

Reviewer: ```suggestion
This document describes the attributes of standalone Events that are represented by `LogRecord`s. All standalone Events have a name and a domain. The Event domain is a namespace for event names and is used as a mechanism to avoid conflicts of event names.
```

**type:** `event`

**Description:** Event attributes.

> 

Reviewer 2: Maybe a section `## Attributes` here?

Author: This was prior to the semconv tool was used to generate the table. I will skip adding a new section as I dont see it other semconv md files too.

> | Attribute  | Type | Description  | Examples  | Required |

Reviewer 2: I believe you can generate this table using the semconv generator tool. 

You need to add 
```
<!-- semconv events-->
<!-- endsemconv -->
```
And the tool will add the table in between for you. 

For an example, you can check this one I did [YAML file](https://github.com/open-telemetry/opentelemetry-specification/blob/main/semantic_conventions/trace/cloudevents.yaml), [Markdown](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/cloudevents.md?plain=1)

Reviewer 2: If something doesn't work, then we need to raise an issue on the generator repo, so we can enhance it further to support this.

Author: Thanks for let me know about this. It works and I made the changes.

|---|---|---|---|---|

> | `event.name` | string | Name or type of the event. | `network-change`; `button-click`; `exception` | Yes |

Reviewer: We no longer use "Yes/No", there are now 4 different requirement levels for semantic conventions: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-requirement-level.md

Author: I made the changes to auto-generate this content from the yaml, which uses the requirement level field.

# Semantic Convention for event attributes

**Status**: [Experimental](../../document-status.md)

This document describes the attributes of standalone Events that are represented
by `LogRecord`s. All standalone Events have a name and a domain.The Event domain
is a namespace for event names and is used as a mechanism to avoid conflicts of
event names.

<!-- semconv event -->
| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|
| `event.name` | string | The name identifies the event. | `click`; `exception` | Required |

> | `event.domain` | string | The domain identifies the context in which an event happened. An event name is unique only within a domain. [1] | `browser`; `mobile`; `kubernetes` | Recommended |

Reviewer 2: Shouldn't we make the domain required? then this sentence can be removed, right? 

>No claim is made about the uniqueness of `event.name`s in the absence of `event.domain`.

Author: There isn't a strong reason why it is not required. Since this field is new and specific to OTel, we were not sure if people will always have a value for it.

Btw, based on a discussion last week, I am going to change this to an enum specifying a few suggested values and allowing custom values, however the requirement level is still going to be `Recommended`.

Reviewer 3: I'm in favor of making it required.

The reason its not required now is probably a reflection of it not being enforceable as required in the API design: You obtain a logger from a logger provider. Like tracer and meter, obtaining a logger requires a name and has optional schema url, version, and scope attributes. `event.domain` is expected to be provided as one of the optional scope attributes. So you can use the API to obtain a logger and emit an event without having included the domain. However, this is outside the expected usage of the API and definitely not something we want to promote.

If we require `event.domain` here, we're saying that any event emitted from the API where a `event.domain` wasn't included as a scope attribute isn't actually an event and is actually just an ordinary log record. That seems pretty reasonable to me.

Author: I have made the changes suggested here.

# Semantic Convention for event attributes

**Status**: [Experimental](../../document-status.md)

This document describes the attributes of standalone Events that are represented

> by `LogRecord`s. All standalone Events have a name and a domain.The Event domain

Reviewer: ```suggestion
by `LogRecord`s. All standalone Events have a name and a domain. The Event domain
```

is a namespace for event names and is used as a mechanism to avoid conflicts of
event names.

<!-- semconv event -->
| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|
| `event.name` | string | The name identifies the event. | `click`; `exception` | Required |

> | `event.domain` | string | The domain identifies the context in which an event happened. An event name is unique only within a domain. [1] | `browser` | Required |

Reviewer 4: As we have been discussion in the RUM Sig, as part of the "general" shape of an event this specification (I believe) should include the `event.data` attribute name which will be where events include their payload.

Author: We will bring this conversation back once https://github.com/open-telemetry/opentelemetry-specification/issues/376 is resolved.

> | `event.domain` | string | The domain identifies the context in which an event happened. An event name is unique only within a domain. [1] | `browser` | Required |

Reviewer 5: Should we also define `event.id | int`? Doesn't have to be required, but I know of large teams where unique well-defined event identifiers are a hard requirement.

/cc @reyang

Author: I only added those semantic conventions required for the API. `event.id` could be added separately later.

**[1]:** An `event.name` is supposed to be unique only in the context of an
`event.domain`, so this allows for two events in different domains to
have same `event.name`, yet be unrelated events.

`event.domain` has the following list of well-known values. If one of them applies, then the respective value MUST be used, otherwise a custom value MAY be used.

| Value  | Description |
|---|---|

> | `browser` | Evens from browser apps |

Reviewer 6: ```suggestion
| `browser` | Events from browser apps |
```

Reviewer 2: I think this is still wrong, no? @scheler

# Semantic Convention for event attributes

**Status**: [Experimental](../../document-status.md)

This document describes the attributes of standalone Events that are represented
by `LogRecord`s. All standalone Events have a name and a domain.The Event domain
is a namespace for event names and is used as a mechanism to avoid conflicts of
event names.

<!-- semconv event -->
| Attribute  | Type | Description  | Examples  | Requirement Level |
|---|---|---|---|---|
| `event.name` | string | The name identifies the event. | `click`; `exception` | Required |
| `event.domain` | string | The domain identifies the context in which an event happened. An event name is unique only within a domain. [1] | `browser` | Recommended |

**[1]:** An `event.name` is supposed to be unique only in the context of an
`event.domain`, so this allows for two events in different domains to
have same `event.name`, yet be unrelated events. No claim is made
about the uniqueness of `event.name`s in the absence of `event.domain`.

> `event.domain` has the following list of well-known values. If one of them applies, then the respective value MUST be used, otherwise a custom value MAY be used.

Reviewer: I spent a bit time thinking about event.domain value and I think we can do this:
1. event.domain values are allocated by Otel spec (can be listed in a table in this doc as it is already done).
2. For relevant domains the event.domain value is equal to the namespace of the existing semantic conventions, e.g. `browser` is the event.domain for browser events and `browser.` is the namespace for browser spans, similarly `k8s` is the domain for Kubernetes events.
3. Addition of new event.domain values is done by new PRs to spec. Similar to addition of new namespaces in semantic conventions.
4. Company-specific event domains start with reverse FQDN. App-specific domains can use company's internal process. This is same as we have for [attribute names](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-naming.md#recommendations-for-application-developers).

Author: @tigrannajaryan clarified that the `device.` prefix would refer only to mobile. For server side devices, `hw.` could be used.

| Value  | Description |
|---|---|
| `browser` | Evens from browser apps |

> | `mobile` | Events from mobile apps |

Reviewer: If we go with my suggestion in the previous comment I think this should be `device` instead of `mobile`.

Reviewer 7: @tigrannajaryan Why device? Are you thinking that domain names should map to an existing set of resource attributes? If the `device` namespace can cover other devices (e.g. IoT), then some mobile-specific events will not be relevant to all devices.

Btw, I have drafted a [PR for sem conventions for two browser events](https://github.com/martinkuba/opentelemetry-specification/pull/1), which shows a possible structure in the semantic conventions spec.

Author: We will go with `device` namespace for mobile specific events as it has already been in the semantic conventions. If there are other events for IoT scenario that are not applicable to mobile, then we will have to pick another namespace for IoT.

> | `kubernetes` | Events from Kubernetes |

Reviewer: If we go with my suggestion in the previous comment I think this should be `k8s` instead of `kubernetes`.
